frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/ApplicationMaster.java

```diff
@@ -344,18 +347,42 @@ private ContainerRequest setupContainerRequest(TaskStatus taskStatus) throws Exc
           requestResource, maxResource);
     }
 
-    if (requestResource.getGpuNumber() > 0) {
-      updateNodeReports(yarnClient.getNodeReports(NodeState.RUNNING));
+    ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
```

Comment:
can you add a sample json file in unit test and demonstrate the behavior when port number is specified, not specified, specified incorrectly?

Reply:
JSon file parsing/validation is the duty of Launcher API. the backend UT can only cover the scenarios  portNumber, portRange different setting in SelectionManagerTest

---
In reply to: [166912476](https://github.com/Microsoft/pai/pull/318#discussion_r166912476) [](ancestors = 166912476)